perception/ground_segmentation/src/scan_ground_filter_nodelet.cpp

```diff
@@ -91,6 +163,236 @@ void ScanGroundFilterComponent::calcVirtualGroundOrigin(pcl::PointXYZ & point)
+{
+  float tmp_delta_max_z = p.orig_point->z - gnd_grids_list.back().max_height;
+  float tmp_delta_avg_z = p.orig_point->z - gnd_grids_list.back().avg_height;
+  float tmp_delta_radius = p.radius - gnd_grids_list.back().radius;
+  float local_slope = std::atan(tmp_delta_avg_z / tmp_delta_radius);
+
+  if (
+    abs(local_slope) < local_slope_max_angle_rad_ ||
+    abs(tmp_delta_avg_z) < non_ground_height_threshold_ ||
+    abs(tmp_delta_max_z) < non_ground_height_threshold_) {
+    gnd_cluster.addPoint(p.radius, p.orig_point->z);
+    p.point_state = PointLabel::GROUND;
+  } else if (local_slope > global_slope_max_angle_rad_) {
+    p.point_state = PointLabel::NON_GROUND;
+  }
+}
+
+void ScanGroundFilterComponent::checkBreakGndGrid(
+  PointRef & p, const std::vector<GridCenter> & gnd_grids_list, PointsCentroid & gnd_cluster)
+{
+  float tmp_delta_avg_z = p.orig_point->z - gnd_grids_list.back().avg_height;
+  float tmp_delta_radius = p.radius - gnd_grids_list.back().radius;
+  float local_slope = std::atan(tmp_delta_avg_z / tmp_delta_radius);
+  if (abs(local_slope) < global_slope_max_angle_rad_) {
+    gnd_cluster.addPoint(p.radius, p.orig_point->z);
+    p.point_state = PointLabel::GROUND;
+  } else if (local_slope > global_slope_max_angle_rad_) {
+    p.point_state = PointLabel::NON_GROUND;
+  }
+}
+void ScanGroundFilterComponent::classifyPointCloudGridScan(
+  std::vector<PointCloudRefVector> & in_radial_ordered_clouds,
+  pcl::PointIndices & out_no_ground_indices)
+{
+  out_no_ground_indices.indices.clear();
+  for (size_t i = 0; i < in_radial_ordered_clouds.size(); i++) {
+    PointsCentroid ground_cluster;
+    ground_cluster.initialize();
+    std::vector<GridCenter> gnd_grids;
+    GridCenter curr_gnd_grid;
+
+    // check empty ray
+    if (in_radial_ordered_clouds[i].size() == 0) {
+      continue;
+    }
+
+    // check the first point in ray
+    auto * p = &in_radial_ordered_clouds[i][0];
+    PointRef * prev_p;
+    prev_p = &in_radial_ordered_clouds[i][0];  // for checking the distance to prev point
+
+    bool init_flg = false;
+    bool prev_list_init = false;
+
+    for (size_t j = 0; j < in_radial_ordered_clouds[i].size(); j++) {
+      p = &in_radial_ordered_clouds[i][j];
+      float global_slope_p = std::atan(p->orig_point->z / p->radius);
+      float non_ground_height_threshold_local = non_ground_height_threshold_;
+      if (p->orig_point->x < -less_interest_dist_) {
+        non_ground_height_threshold_local =
+          non_ground_height_threshold_ * (p->radius / less_interest_dist_);
+      }
+      // classify first grid's point cloud
+      if (
+        !init_flg && p->radius <= first_ring_distance_ &&
+        global_slope_p >= global_slope_max_angle_rad_ &&
+        p->orig_point->z > non_ground_height_threshold_local) {
+        out_no_ground_indices.indices.push_back(p->orig_index);
+        p->point_state = PointLabel::NON_GROUND;
+        prev_p = p;
+        continue;
+      }
```

Comment:
[nits] It would be more readable if it could be functionalized in this block.
i.g.  
```
      if (isFisrtPointAboveGlobalThreshold(some, necessary, variables)){
        out_no_ground_indices.indices.push_back(p->orig_index);
        p->point_state = PointLabel::NON_GROUND;
        prev_p = p;
        continue;
      }
```